refactor/version-1

0xm1kr opened the pull request:
This PR updates Ceramic dependencies to support the updated Ceramic Streams. This is an intermediary update primarily for testing purposes while work on https://github.com/3box/3box/issues/1401 is in progress. The package was tested with a small test environment which appears to work as expected (as long as the NFT doesn't change owners).

**Changes**

* Update ceramic and peer dependencies to new version 1
    * Switch to using `stream-caip10-link` instead of using the deprecated `ceramic.createDocument`
* Update / fix tests
    * Update to use `jest-environment-ceramic` from `ceramicstudio/js-idx`
* Add types generation and support for typescript npm module
* Fix some typings and misspellings

Commit:
chore(deps): update ceramic to v1

Commit:
chore: update to make use of stream-caip10-link

Commit:
chore: fix types and tests

stbrody commented on src/index.ts:
```diff
@@ -73,39 +74,22 @@ async function assetToAccount(
 async function accountsToDids(
   accounts: AccountID[], 
-  timestamp: number, 
+  // NOTE: Opts: { atTime: timestamp } no longer applicable for create?
```
Can you elaborate on the issue you saw trying to use atTime with Caip10Link?  What code did you try and what error did you see?

0xm1kr replied:
> Can you elaborate on the issue you saw trying to use atTime with Caip10Link? What code did you try and what error did you see?

I'm using `Caip10Link.fromAccount()` from `@ceramicnetwork/stream-caip10-link` ( version ^1.0.6 ).

It may just be a type error but it appears as though the `CreateOpts` type extends `BasicLoadOpts` instead of `LoadOpts` which does not contain the `atTime` property.

```
interface BasicLoadOpts {
    sync?: SyncOptions;
    syncTimeoutSeconds?: number;
}
export interface LoadOpts extends BasicLoadOpts {
    atTime?: number;
}
export interface UpdateOpts {
    anchor?: boolean;
    publish?: boolean;
}
export interface CreateOpts extends UpdateOpts, BasicLoadOpts {
}
```

stbrody replied:
Ah yes, sorry about that, that seems like an oversight. Fix is here: https://github.com/ceramicnetwork/js-ceramic/pull/1495

0xm1kr replied:
Oh nice, no worries. Thanks for jumping on it!

mikehroth commented:
Oh nice, no worries. Thanks for jumping on it!

On Mon, Jun 21, 2021 at 2:45 PM Spencer T Brody ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/index.ts
> <https://github.com/ceramicnetwork/nft-did-resolver/pull/4#discussion_r655691911>
> :
>
> >  /**
>   * Creates CAIP-10 links for each account to be used as controllers.
>   * Since there may be many owners for a given NFT (only ERC1155 for now), there can be many
>   * controllers of that DID document.
>   */
>  async function accountsToDids(
>    accounts: AccountID[],
> -  timestamp: number,
> +  // NOTE: Opts: { atTime: timestamp } no longer applicable for create?
>
> Ah yes, sorry about that, that seems like an oversight. Fix is here:
> ceramicnetwork/js-ceramic#1495
> <https://github.com/ceramicnetwork/js-ceramic/pull/1495>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ceramicnetwork/nft-did-resolver/pull/4#discussion_r655691911>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AOP354O2I5ZXPOYZAOCSSRLTT6QEXANCNFSM467RNKRQ>
> .
>

oed commented on src/__tests__/index.test.ts:
```diff
@@ -447,11 +452,16 @@ function expectBlockQueries(versionTime: string) {
-  await doc.change({ content: proof });
+  const accountId = await ethAuthProv.accountId()
+  const link = await Caip10Link.fromAccount((global as any).ceramic, accountId, {})
+  // NOTE: is this correct?
+  await link.setDid(caipLinkControllerDid, ethAuthProv)
+  
+  // const proof = await ethAuthProv.createLink(caipLinkControllerDid);
+  // const doc = await ceramic.createDocument('caip10-link', {
+  //   metadata: { family: 'caip10-link', controllers: [proof.account] }
+  // });
+  // await doc.change({ content: proof });
```
Yes this looks correct to me.
```suggestion
  await link.setDid(caipLinkControllerDid, ethAuthProv)
```

oed reviewed (commented):
Looks good.

> appears to work as expected (as long as the NFT doesn't change owners).

I assume this is because we are not using the `atTime` option yet.

0xm1kr committed:
chore: update src/__tests__/index.test.ts

remove comments

Co-authored-by: Joel Thorstensson <oed@3box.io>

Commit:
chore(deps): update to 1.0.7 and reimplement atTime

0xm1kr commented:
@oed @stbrody ready for review/merge 👍

Commit:
chore: update tsconfig to match js-ceramic

oed commented on src/__tests__/index.test.ts:
```diff
@@ -37,8 +38,12 @@ enum ErcNamespace {
   ERC1155 = 'erc1155'
 }
 
+// since v1, tests require more time updated
+// to 12s timeout to prevent test timeouts
+jest.setTimeout(12000);
```
This is quite strange. Which tests take 12s?

oed reviewed (commented):
Looks good. Curious what is the cause of the increased time the test takes. Ideally this shouldn't happen. 

Will also try to run these tests locally.

stbrody commented:
> Looks good. Curious what is the cause of the increased time the test takes. Ideally this shouldn't happen.
> 
> Will also try to run these tests locally.

I bet this is due to some of the changed behaviors around default sync time.  For instance now `Caip10Link.fromAccount` will try to sync for 3 seconds by default, as will creating a deterministic TileDocument.

0xm1kr commented:
> > Looks good. Curious what is the cause of the increased time the test takes. Ideally this shouldn't happen.
> > Will also try to run these tests locally.
> 
> I bet this is due to some of the changed behaviors around default sync time. For instance now `Caip10Link.fromAccount` will try to sync for 3 seconds by default, as will creating a deterministic TileDocument.

I think @stbrody may be right. Tests which involve a Caip10Link appear to take at least 3s. 1155s which likely have multiple owners take some multiplier of that time.

Here are the execution times of the tests on my machine:

```
    ERC721 NFTs
      ✓ resolves an erc721 nft document without caip10-link (3035 ms)
      ✓ resolves an erc721 nft document with caip10-link (4 ms)
      ✓ resolves an erc721 nft doc with an older timestamp (5 ms)
      ✓ throws on invalid ERC721 contract (1 ms)
      ✓ throws on non-existent ERC721 token with valid contract (1 ms)
      ✓ throws when an invalid erc namespace is provided in DID
      ✓ resolves erc721 dids with a custom subgraph url (4 ms)
      ✓ allows for erc721 namespace on caip2 chains beside eth (3035 ms)
    ERC1155 NFTs
      ✓ resolves an erc1155 nft document without caip10-link (9107 ms)
      ✓ resolves an erc1155 nft document with caip10-link (10 ms)
      ✓ resolves an erc1155 nft doc with an older timestamp (11 ms)
      ✓ throws on invalid ERC1155 contract
      ✓ throws on non-existent ERC1155 token with valid contract (1 ms)
      ✓ resolves erc1155 dids with a custom subgraph url (10 ms)
      ✓ allows for erc1155 namespace on caip2 chains beside eth (9139 ms)
 ```

stbrody commented:
If you are using `Caip10Link.fromAccount` to create a brand new Caip10Link, rather than to load an existing one, then you can add the option `syncTimeoutSeconds:0` to disable the syncing behavior, which should speed things back up.  This probably only makes sense to do within the tests themselves where you control the full state of the world.  In the actual resolver itself you never know if the Caip10Link may have already been created by someone else, so you probably always want to actually sync there

0xm1kr commented:
> If you are using `Caip10Link.fromAccount` to create a brand new Caip10Link, rather than to load an existing one, then you can add the option `syncTimeoutSeconds:0` to disable the syncing behavior, which should speed things back up. This probably only makes sense to do within the tests themselves where you control the full state of the world. In the actual resolver itself you never know if the Caip10Link may have already been created by someone else, so you probably always want to actually sync there

This is a great suggestion for the tests, however, it would probably require adding a property of some kind to the `NftResolverConfig` to bypass the sync. I could see this being useful under circumstances where a user might know for a fact that the `Caip10Link`s have already been created? Let me know if you and @oed think its worth it to add the property to the config and I can adjust then remove the updated `jestTimeout`.

stbrody commented on src/__tests__/index.test.ts:
```diff
@@ -447,11 +452,9 @@ function expectBlockQueries(versionTime: string) {
-  });
-  await doc.change({ content: proof });
+  const accountId = await ethAuthProv.accountId()
+  const link = await Caip10Link.fromAccount((global as any).ceramic, accountId, {})
```
```suggestion
  const link = await Caip10Link.fromAccount((global as any).ceramic, accountId, {syncTimeoutSeconds: 0})
```

stbrody replied:
Does this change speed up the tests at all?

0xm1kr replied:
Unfortunately no, that only speeds up the initial Caip10Link for the `EthAuthProvider` (https://github.com/ManyUses/nft-did-resolver/blob/refactor/version-1/src/__tests__/index.test.ts#L69)

It would need to be added in the resolver itself here:
https://github.com/ManyUses/nft-did-resolver/blob/refactor/version-1/src/index.ts#L90

oed replied:
Yeah, I think this would speed it up. Should be completely fine to do this in the test file.

oed commented:
@mikehroth The only thing I would change in the tests is to create the caip10-links for erc1155 in parallel.

Commit:
chore: create caip10-links async

Commit:
chore: remove jestTimeout adjustment in tests

0xm1kr commented:
@oed / @stbrody based on our conversations:
- Made `Caip10Link` creation async
- Added `{ syncTimeoutSeconds: 0 }` to tests
- Removed jestTimeout change in tests

oed reviewed (approved):
Tests are passing for me locally.
